pkg/azurefile/controllerserver.go

```diff
@@ -986,6 +989,107 @@ func (d *Driver) ListSnapshots(_ context.Context, _ *csi.ListSnapshotsRequest) (
+			return fmt.Errorf("get src accountkey get err: %v", err)
+		}
+		klog.V(2).Infof("generate sas token for resource group(%s) src account(%s)", resourceGroupName, srcAccountName)
+		if srcAccountSasToken, err = generateSASToken(srcAccountName, srcAccountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes); err != nil {
+			return err
+		}
+	}
+
+	srcPath := fmt.Sprintf("https://%s.file.%s/%s%s&sharesnapshot=%s", srcAccountName, storageEndpointSuffix, srcFileShareName, srcAccountSasToken, snapshot)
+	dstPath := fmt.Sprintf("https://%s.file.%s/%s%s", dstAccountName, storageEndpointSuffix, dstFileShareName, dstAccountSasToken)
+
+	return d.copyFileShareByAzcopy(ctx, srcFileShareName, dstFileShareName, srcPath, dstPath, snapshot, srcAccountName, resourceGroupName, srcAccountSasToken, authAzcopyEnv, secretName, secretNamespace, secrets, accountOptions, storageEndpointSuffix)
+}
+
+func (d *Driver) copyFileShareByAzcopy(ctx context.Context, srcFileShareName, dstFileShareName, srcPath, dstPath, snapshot, accountName, resourceGroupName, accountSASToken string, authAzcopyEnv []string, secretName, secretNamespace string, secrets map[string]string, accountOptions *azure.AccountOptions, storageEndpointSuffix string) error {
+	azcopyCopyOptions := defaultAzcopyCopyOptions
+	if snapshot != "" {
+		srcFileShareName = fmt.Sprintf("%s(snapshot: %s)", srcFileShareName, snapshot)
+		azcopyCopyOptions = snapshotAzcopyCopyOptions
+	}
+
+	jobState, percent, err := d.azcopy.GetAzcopyJob(dstFileShareName, authAzcopyEnv)
+	klog.V(2).Infof("azcopy job status: %s, copy percent: %s%%, error: %v", jobState, percent, err)
+	switch jobState {
+	case fileutil.AzcopyJobError, fileutil.AzcopyJobCompleted:
+		return err
+	case fileutil.AzcopyJobRunning:
+		return fmt.Errorf("an existed azcopy job is running, copy percent: %s%%, please wait for it to complete", percent)
+	case fileutil.AzcopyJobNotFound:
+		klog.V(2).Infof("copy fileshare %s to %s", srcFileShareName, dstFileShareName)
+		copyErr := fileutil.WaitForExecCompletion(time.Duration(d.waitForAzCopyTimeoutMinutes)*time.Minute, func() error {
+			if out, err := d.execAzcopyCopy(srcPath, dstPath, azcopyCopyOptions, authAzcopyEnv); err != nil {
+				return fmt.Errorf("exec error: %v, output: %v", err, string(out))
+			}
+			return nil
+		}, func() error {
+			_, percent, _ := d.azcopy.GetAzcopyJob(dstFileShareName, authAzcopyEnv)
+			return fmt.Errorf("timeout waiting for copy fileshare %s to %s succeed, copy percent: %s%%", srcFileShareName, dstFileShareName, percent)
+		})
+		if accountSASToken == "" && copyErr != nil && strings.Contains(copyErr.Error(), authorizationPermissionMismatch) {
+			klog.Warningf("azcopy list failed with AuthorizationPermissionMismatch error, should assign \"Storage File Data SMB Share Elevated Contributor\" role to controller identity, fall back to use sas token, original error: %v", copyErr)
```

Comment:
managed identity does not work now, let's remove this logic until it's supported in azcopy

Reply:
let's remove managed identity code to make it more clear.

Reply:
let's remove managed identity code to make it more clear.